src/cve/utils/source_code_git_loader.py

```diff
@@ -109,27 +109,26 @@ def load_repo(self):
                 if repo.remotes.origin.url != self.clone_url:
                     raise ValueError("A different repository is already cloned at this path.")
 
-                logger.debug("Updating existing Git repo for URL '%s' and branch '%s'", self.clone_url, self.branch)
-
-                # Reliable way to checkout the branch of a shallow clone
-                repo.git.fetch("origin", self.branch, depth=1)
-                repo.git.checkout("FETCH_HEAD")
+                logger.debug("Updating existing Git repo for URL '%s' @ '%s'", self.clone_url, self.ref)
             else:
-                logger.debug("Cloning repository from URL: '%s' for branch/tag '%s'", self.clone_url, self.branch)
-
-                repo = Repo.clone_from(self.clone_url, self.repo_path, branch=self.branch, depth=1)
+                logger.debug("Cloning repository from URL: '%s' @ '%s'", self.clone_url, self.ref)
+                # Create a shallow clone of the repository without checking out
+                repo = Repo.clone_from(self.clone_url, self.repo_path, depth=1, no_checkout=True)
 
-                # When creating the directory, ensure its a safe directory
+                # Set repo as git safe directory to avoid errors if directory ownership is changed outside the pipeline
+                # https://git-scm.com/docs/git-config#Documentation/git-config.txt-safedirectory
                 with repo.config_writer(config_level="global") as config:
                     config.add_value("safe", "directory", str(self.repo_path.absolute()))
 
         else:
-            logger.debug("Using existing Git repo at path: '%s' for branch/tag '%s'", self.repo_path, self.branch)
-
             repo = Repo(self.repo_path)
-            repo.git.checkout(self.branch)
+            logger.debug("Using existing Git repo at path: '%s' @ '%s'", self.repo_path, self.ref)
+
+       # Reliable way to check out the ref using a shallow clone
+        repo.git.fetch("origin", self.ref, depth=1)
+        repo.git.checkout("FETCH_HEAD")
```

Comment:
Minor fixup: need one more space in the comment
```suggestion
        # Reliable way to check out the ref using a shallow clone
        repo.git.fetch("origin", self.ref, depth=1)
        repo.git.checkout("FETCH_HEAD")
```